### PR TITLE
Fix building TupleDomain for comparisons with cercions (v2)

### PR DESCRIPTION
**Pull request by losipiuk**
supersedes https://github.com/prestodb/presto/pull/7102

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 1]
```diff
@@ -692,34 +689,101 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
...
+                symbolExpression = ((Cast) symbolExpression).getExpression();
+            }
+            else {
+                return false;
+            }
         }
-        return Optional.empty();
+    }
+
+    private static boolean isCoercion(Metadata metadata, Map<Expression, Type> expressionTypes, Cast cast)
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: to be more specific, this should be called isImplicitCoercion

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 2]
```diff
@@ -692,34 +689,101 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
...
+    }
+
+    private static boolean checkSymbolExpressionValid(Metadata metadata, IdentityHashMap<Expression, Type> expressionTypes, Expression symbolExpression)
+    {
+        while (true) {
+            if (symbolExpression instanceof SymbolReference) {
+                return true;
+            }
+            else if (symbolExpression instanceof Cast) {
+                if (!isCoercion(metadata, expressionTypes, (Cast) symbolExpression)) {
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: After thinking about this in more depth, I don't think we can handle this operation in this fashion, even with only unrolling implicit coercions.

For example, I think you mentioned this example before, but with the case
```
CAST(my_bigint AS DOUBLE) = 1.1
```
This is a valid implicit coercion, but it is not equivalent to my_bigint = 1.

We may be able to make this work if we can succinctly find a way to define how implicit coercions operate. For example, we may need to handle the different types of comparisons in different ways (sort of like how rewriteComparisonExpression does that).

Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: Actually this case should work fine.
Tests for cases like that (though not exactly BIGINT vs DOUBLE) are in testNumericTypeTranslation.

Such expression would be rewritten to expression which is false for all non-null values in `my_bigint`.
Because `1.1 != cast(floor_cast(1.1))`.

Take a look at section:
```
case EQUAL: {
                    if (coercedValueIsEqualToOriginal) {
                        return new ComparisonExpression(EQUAL, symbolExpression, coercedLiteral);
                    }
                    // Return something that is false for all non-null values
                    return and(new ComparisonExpression(EQUAL, symbolExpression, coercedLiteral),
                            new ComparisonExpression(NOT_EQUAL, symbolExpression, coercedLiteral));
                }
```
in `rewriteComparisonExpression`.

Actually I am not sure if we should keep `NULL` values here. Or rather return empty domain for such case.
Is TupleDomain used in contexts where we want to distinguish `NULL` and `FALSE` as result from predicate?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 1 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: I will change it as soon as I get to computer with source code.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java [hunk 3]
```diff
@@ -40,6 +40,7 @@
 import com.facebook.presto.sql.tree.NullLiteral;
 import com.facebook.presto.sql.tree.QualifiedName;
 import com.facebook.presto.sql.tree.StringLiteral;
+import com.facebook.presto.type.TypeUtils;
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java]: make tests refactoring a separate commit?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java [hunk 4]
```diff
@@ -777,6 +774,21 @@ public void testFromBasicComparisonsWithNulls()
     }
 
     @Test
+    void testTimestampToDateExplicitCastOnSymbol()
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDomainTranslator.java]: add test for other way around too? date to timestamp?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 5]
```diff
@@ -692,34 +689,101 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
...
-
         Preconditions.checkArgument(leftType.equals(rightType), "left and right type do not match in comparison expression (%s)", comparison);
 
-        if (left instanceof SymbolReference && !(right instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) left, comparison.getType(), new NullableValue(rightType, right)));
+        if (left instanceof Expression && right instanceof Expression) {
+            return Optional.empty();
+        }
+
+        if (!(left instanceof Expression) && !(right instanceof Expression)) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: isn't that technically illegal argument too? comparison should fold to constant if both sides are evaluated

Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: I think this is case we need to handle her. As It is possible that one of the optimisers calls `DomainTranslator` for expression like
``` ... WHERE 3 = 5```.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 1 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 6]
```diff
@@ -689,31 +687,103 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
...
+                    // example which illustrates the problem:
+                    //
+                    // let t be of timestamp type:
+                    //
+                    // and expression be:
+                    // cast(t as date) == date_literal
+                    //
+                    // after dropping cast we end up with:
+                    //
+                    // t == date_literal
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: Don't know if it make sense, but i would add here an example with DECIMAL that we have found as well.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 7]
```diff
@@ -689,31 +687,103 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
...
+        return Optional.of(new NormalizedSimpleComparison(symbolExpression, comparisonType, value));
+    }
+
+    private static Type typeOf(Expression expression, Session session, Metadata metadata, Map<Symbol, Type> types)
+    {
+        IdentityHashMap<Expression, Type> expressionTypes = ExpressionAnalyzer.getExpressionTypes(session, metadata, new SqlParser(), types, expression, emptyList() /* parameters already replaced */);
+        return expressionTypes.get(expression);
+    }
+
+    private static boolean checkSymbolExpressionValid(Metadata metadata, IdentityHashMap<Expression, Type> expressionTypes, Expression symbolExpression)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: Maybe change the name to better express what this method does. What it actually does - it just checks whether the cast chain is strictly widening. Maybe call it `downcastsPresent` and invert the result?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 8]
```diff
@@ -397,36 +399,32 @@ protected ExtractionResult visitComparisonExpression(ComparisonExpression node,
...
-                return createComparisonExtractionResult(normalized.getComparisonType(), symbol, fieldType, coercedValue.get().getValue(), complement);
+            Expression symbolExpression = normalized.getSymbolExpression();
+            if (symbolExpression instanceof SymbolReference) {
+                Symbol symbol = Symbol.from(symbolExpression);
+                NullableValue value = normalized.getValue();
+                Type type = value.getType(); // common type for symbol and value
+                return createComparisonExtractionResult(normalized.getComparisonType(), symbol, type, value.getValue(), complement);
             }
+            else if (symbolExpression instanceof Cast) {
+                Cast castExpression = (Cast) symbolExpression;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: Cast is being un-wrapped only once. Aren't we suppose to unwrap cast until we actually reach a symbol reference? If we assume that double coercion will be simplified by the optimizer then shouldn't we add explicit check here? To make sure that after single unwrap we actually get the symbol reference? Also if we assume so, the `checkSymbolExpressionValid` must be changed. So it won't traverse deeper that 1 level deep?

Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: NVM. I see, you are doing it recursively by running `process(coercedExpression.get(), complement)`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 9]
```diff
@@ -397,36 +399,32 @@ protected ExtractionResult visitComparisonExpression(ComparisonExpression node,
...
+                NullableValue value = normalized.getValue();
+                Type valueType = value.getType(); // type of value
+                Type castSourceType = typeOf(castExpression.getExpression(), session, metadata, types); // type of expression which is then cast to type of value
+
+                // we use saturated floor cast value -> castSourceType to rewrite original expression to new one with one cast peeled off the symbol side
+                Optional<Expression> coercedExpression = coerceComparisonWithRounding(
+                        castSourceType, castExpression.getExpression(), valueType, value.getValue(), normalized.getComparisonType());
+
+                if (coercedExpression.isPresent()) {
+                    return process(coercedExpression.get(), complement);
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: Can we add some test cases that verify the case of double/triple unwrap?

**Comment by arhimondr** [comment]
LGTM % Nits

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 9 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 7 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 6 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: I will add test for that

**Comment by losipiuk** [comment]
@erichwang I addressed all the comments and replied to your concerns.
Please take a look when you have a chance.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 10]
```diff
@@ -689,31 +687,103 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
         Object left = ExpressionInterpreter.expressionOptimizer(comparison.getLeft(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
         Object right = ExpressionInterpreter.expressionOptimizer(comparison.getRight(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
 
-        if (left instanceof SymbolReference && !(right instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) left, comparison.getType(), new NullableValue(expressionTypes.get(comparison.getRight()), right)));
+        Type leftType = expressionTypes.get(comparison.getLeft());
+        Type rightType = expressionTypes.get(comparison.getRight());
+        Preconditions.checkArgument(leftType.equals(rightType), "left and right type do not match in comparison expression (%s)", comparison);
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: static import checkArgument

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 11]
```diff
@@ -689,31 +687,103 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
...
+
+        if (left instanceof Expression && right instanceof Expression) {
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: you can just do:
```java
left instanceof Expression == right instanceof Expression
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 12]
```diff
@@ -689,31 +687,103 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
...
+            comparisonType = comparison.getType();
+            value = new NullableValue(rightType, right);
+        }
+        else {
+            symbolExpression = comparison.getRight();
+            comparisonType = comparison.getType().flip();
+            value = new NullableValue(leftType, left);
+        }
+
+        if (!checkOnlyImplicitCastsAroundSymbol(metadata, expressionTypes, symbolExpression)) {
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: instead of checking the implicit casts here and having that information be retained in the NormalizedSimpleComparison by convention, it seems cleaner just to return the arbitrary Expression in NormalizedSimpleComparison, and have the calling code do the implicit coercion check as it is unpeeling the casts.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 13]
```diff
@@ -689,31 +687,103 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
...
+        return Optional.of(new NormalizedSimpleComparison(symbolExpression, comparisonType, value));
+    }
+
+    private static Type typeOf(Expression expression, Session session, Metadata metadata, Map<Symbol, Type> types)
+    {
+        IdentityHashMap<Expression, Type> expressionTypes = ExpressionAnalyzer.getExpressionTypes(session, metadata, new SqlParser(), types, expression, emptyList() /* parameters already replaced */);
+        return expressionTypes.get(expression);
+    }
+
+    private static boolean checkOnlyImplicitCastsAroundSymbol(Metadata metadata, IdentityHashMap<Expression, Type> expressionTypes, Expression symbolExpression)
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: do we need to make these above two functions static? it seems like it might be cleaner if we only passed in the expressions.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 10 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 11 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 13 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: It is called from toNormalizedSimpleComparison which was static already.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 12 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: I am not sure if that would be cleaner. We would need to have expressionTypes (so perfome analysis on expression) in `visitComparisonExpression`.
I will try to refactor it.

**Comment by losipiuk** [comment]
@erichwang I refactored the code as you suggested. It seems nicer.
I left fixup commit for refactoring - please squash it if you are going to merge.

I removed the "Fix index error in FunctionInvoker when accessing isNull array" as it this bug not seem to be triggered by this code anymore after changes. I will prepare another review for that one.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java [hunk 14]
```diff
@@ -391,42 +392,107 @@ protected ExtractionResult visitNotExpression(NotExpression node, Boolean comple
...
-            NullableValue value = normalized.getValue();
-
-            Optional<NullableValue> coercedValue = coerce(value, fieldType);
-            if (coercedValue.isPresent()) {
-                return createComparisonExtractionResult(normalized.getComparisonType(), symbol, fieldType, coercedValue.get().getValue(), complement);
+            Expression symbolExpression = normalized.getSymbolExpression();
+            if (symbolExpression instanceof SymbolReference) {
+                Symbol symbol = Symbol.from(symbolExpression);
+                NullableValue value = normalized.getValue();
+                Type type = value.getType(); // common type for symbol and value
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: if it's easy, we should just add a verify here to assert that they are the same types. if not, don't worry about it.

**Comment by erichwang** [comment]
Also, please merge your fixup.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java (hunk 14 above)
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java]: Currently type of expression is derived internally in `toNormalizedSimpleComparison`. And that code does the assertion that both sides of comparison have same type.
I could do this here again. Or do it only here and pass `Expression->Type` map to `toNormalizedSimpleComparison` but I am not sure if that would actually be improvement. 

I would leave it as it is. IMO it is good enough.

**Comment by losipiuk** [comment]
> Just one tiny comment left, otherwise this looks good. I'll wait until this next release goes out before merging.

Thanks for review! I rebased and squashed the fixup.

**Comment by erichwang** [comment]
Merged, thanks!
